graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/nodes/promise/ImportCallNode.java

```diff
@@ -185,24 +221,37 @@ public DynamicObject createImportModuleDynamicallyJob(ScriptOrModule referencing
         if (context.isOptionTopLevelAwait()) {
             Triple<ScriptOrModule, String, PromiseCapabilityRecord> request = new Triple<>(referencingScriptOrModule, specifier, promiseCapability);
             PromiseCapabilityRecord startModuleLoadCapability = newPromiseCapability();
-            PromiseReactionRecord startModuleLoad = PromiseReactionRecord.create(startModuleLoadCapability, createImportModuleDynamicallyHandler(), true);
+            PromiseReactionRecord startModuleLoad = PromiseReactionRecord.create(startModuleLoadCapability, createImportModuleDynamicallyHandler(RootType.TLA), true);
             return promiseReactionJobNode.execute(startModuleLoad, request);
         } else {
             Pair<ScriptOrModule, String> request = new Pair<>(referencingScriptOrModule, specifier);
-            return promiseReactionJobNode.execute(PromiseReactionRecord.create(promiseCapability, createImportModuleDynamicallyHandler(), true), request);
+            return promiseReactionJobNode.execute(PromiseReactionRecord.create(promiseCapability, createImportModuleDynamicallyHandler(RootType.IMD), true), request);
         }
     }
 
+    public DynamicObject createImportModuleDynamicallyJob(ScriptOrModule referencingScriptOrModule, DynamicObject specifier, Source source, PromiseCapabilityRecord promiseCapability) {
+        Triple<ScriptOrModule, DynamicObject, Source> request = new Triple<>(referencingScriptOrModule, specifier, source);
+
+        return promiseReactionJobNode.execute(PromiseReactionRecord.create(promiseCapability, createImportModuleDynamicallyHandler(RootType.MB), true), request);
+    }
+
     /**
      * Returns a handler function to be used together with a PromiseReactionJob in order to perform
      * the steps of both HostImportModuleDynamically and FinishDynamicImport.
      */
-    private DynamicObject createImportModuleDynamicallyHandler() {
-        JSFunctionData functionData = context.getOrCreateBuiltinFunctionData(JSContext.BuiltinFunctionKey.ImportModuleDynamically, (c) -> createImportModuleDynamicallyHandlerImpl(c));
+    private DynamicObject createImportModuleDynamicallyHandler(RootType root) {
+        JSFunctionData functionData = context.getOrCreateBuiltinFunctionData(JSContext.BuiltinFunctionKey.ImportModuleDynamically, (c) -> createImportModuleDynamicallyHandlerImpl(c, root));
         return JSFunction.create(context.getRealm(), functionData);
     }
 
-    private static JSFunctionData createImportModuleDynamicallyHandlerImpl(JSContext context) {
+    private static enum RootType {
+        IMD,
```

Comment:
I would prefer to use full names here rather than an acronym (i.e., `TopLevelAwait`, `ModuleBlock` etc)

Reply:
Refactored to full names.